File: src/agentcap/drivers/goose.py
"""Goose driver.

Drives ``goose run -t "<prompt>"`` non-interactively. The proxy URL +
provider + ``OPENAI_API_KEY`` are baked into the per-agent image's
ENV (see [containers/agentcap-goose.Containerfile](
../../../containers/agentcap-goose.Containerfile)); the driver only
sets ``GOOSE_MODEL`` per run.

Goose's own session state lives at ``~/.config/goose/sessions/``
inside the sandbox, redirected to the bind-mounted ``state/`` dir
so it survives ``podman run --rm`` boundaries between turns.
"""

from __future__ import annotations

import subprocess
import uuid
from pathlib import Path
from typing import Sequence

from . import AgentDriver, AgentTurn
from ..sandbox import Sandbox


def parse_tool_errors(stdout: str) -> list[str]:
    # TODO: goose's tool-error format is not yet characterised.
    return []


class GooseDriver(AgentDriver):
    name = "goose"

    def __init__(
        self,
        *,
        sandbox: Sandbox,
        binary: str = "goose",
        model: str | None = None,
        cwd: Path | str | None = None,
        extra_args: Sequence[str] = (),
    ) -> None:
        self.sandbox = sandbox
        self.binary = binary
        self.model = model
        # ``cwd`` is sandbox-side: a host path bind-mounted into the
        # container at the same path.
        self.cwd = str(cwd) if cwd is not None else None
        self.extra_args = list(extra_args)

    def close(self) -> None:
        """No-op."""

    def _build_argv(
        self, prompt: str, *, session_name: str | None, resume: bool
    ) -> list[str]:
        argv = [self.binary, "run", "-t", prompt, *self.extra_args]
        if session_name is None:
            argv.append("--no-session")
        else:
            argv.extend(["--name", session_name])
            if resume:
                argv.append("--resume")
        return argv

    def _run(
        self,
        argv: list[str],
        env: dict | None,
        timeout: float | None,
    ) -> subprocess.CompletedProcess:
        full_env: dict[str, str] = {}
        if self.model:
            full_env["GOOSE_MODEL"] = self.model
        if env:
            full_env.update(env)
        return self.sandbox.run(
            argv,
            env=full_env,
            cwd=self.cwd,
            timeout=timeout,
        )

    def start(
        self,
        prompt: str,
        *,
        env: dict | None = None,
        timeout: float | None = None,
    ) -> AgentTurn:
        session_name = f"agentcap-{uuid.uuid4().hex[:8]}"
        proc = self._run(
            self._build_argv(prompt, session_name=session_name, resume=False),
            env,
            timeout,
        )
        return AgentTurn(
            session_id=session_name,
            response_text=proc.stdout.strip(),
            returncode=proc.returncode,
            stdout=proc.stdout,
            stderr=proc.stderr,
            tool_errors=parse_tool_errors(proc.stdout),
        )

    def resume(
        self,
        prompt: str,
        *,
        session_id: str,
        env: dict | None = None,
        timeout: float | None = None,
    ) -> AgentTurn:
        proc = self._run(
            self._build_argv(prompt, session_name=session_id, resume=True),
            env,
            timeout,
        )
        return AgentTurn(
            session_id=session_id,
            response_text=proc.stdout.strip(),
            returncode=proc.returncode,
            stdout=proc.stdout,
            stderr=proc.stderr,
            tool_errors=parse_tool_errors(proc.stdout),
        )

File: src/drivers/goose.rs
//! Goose driver: `goose run -t "<prompt>"`. Ports `drivers/goose.py`. The proxy
//! URL + provider are baked into the image ENV; the driver sets `GOOSE_MODEL`.

use std::collections::BTreeMap;
use std::sync::Arc;
use std::time::Duration;

use super::{AgentDriver, AgentTurn, DriverError};
use crate::sandbox::PodmanSandbox;

pub struct GooseDriver {
    sandbox: Arc<PodmanSandbox>,
    cwd: Option<String>,
    model: Option<String>,
}

impl GooseDriver {
    pub fn new(sandbox: Arc<PodmanSandbox>, cwd: Option<String>, model: Option<String>) -> Self {
        GooseDriver { sandbox, cwd, model }
    }

    fn argv(&self, prompt: &str, session_name: &str, resume: bool) -> Vec<String> {
        let mut argv = vec![
            "goose".into(),
            "run".into(),
            "-t".into(),
            prompt.into(),
            "--name".into(),
            session_name.into(),
        ];
        if resume {
            argv.push("--resume".into());
        }
        argv
    }

    fn env(&self) -> BTreeMap<String, String> {
        let mut e = BTreeMap::new();
        if let Some(m) = &self.model {
            e.insert("GOOSE_MODEL".into(), m.clone());
        }
        e
    }

    fn run(&self, argv: &[String], session_name: &str, timeout: Option<Duration>) -> Result<AgentTurn, DriverError> {
        let out = self.sandbox.run(argv, &self.env(), self.cwd.as_deref(), timeout)?;
        Ok(AgentTurn {
            session_id: Some(session_name.to_string()),
            response_text: out.stdout.trim().to_string(),
            returncode: out.code,
            stdout: out.stdout,
            stderr: out.stderr,
            tool_errors: Vec::new(),
        })
    }
}

impl AgentDriver for GooseDriver {
    fn name(&self) -> &str {
        "goose"
    }
    fn start(&self, prompt: &str, timeout: Option<Duration>) -> Result<AgentTurn, DriverError> {
        let session_name = format!("agentcap-{}", &uuid::Uuid::new_v4().simple().to_string()[..8]);
        self.run(&self.argv(prompt, &session_name, false), &session_name, timeout)
    }
    fn resume(&self, prompt: &str, session_id: &str, timeout: Option<Duration>) -> Result<AgentTurn, DriverError> {
        self.run(&self.argv(prompt, session_id, true), session_id, timeout)
    }
}
